text/3594-expose-stackrealign-attribute.md

```diff
@@ -0,0 +1,101 @@
+This alignment is achieved by adjusting the stack pointer accordingly to the stack alignment specified in the target ABI's data layout.
+Adding this attribute unnecessarily might "waste" space on the stack which could be crucial in real-time systems.
+
+In LLVM the `alignstack` gets an argument that specifies the alignment in bytes(also must be a power of 2). 
+Below is an example of how it would work for an example data-layout:
+`e-m:e-i64:64-f80:128-n8:16:32:64-S128`.
+The `S128` is the part that describes the natural stack alignment in bits.
+So practically, we just need to divide that value by 8, and place it as the argument of `alignstack`.
+So in the `S128` case it would look like this: `alignstack=16`.
+
+```
+define i32 @callback_function() #0 {
+start:
+  ret i32 0
+}
+
+attributes #0 = { alignstack=16 }
+```
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+- Introducing a new attribute adds complexity to the language.
+- Limited use cases: Stack realignment may not be necessary in most Rust codebases, potentially making this feature less relevant for many developers.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+An alternative could be a macro workaround instead of adding the attribute.
+However it would be more like band-aid than an actual solution.
+Another alternative could be adding the any extern "C" function the `stackrealign` attribute implicitly which would solve the main use-case.
+An extra option could be not verifying data-layout for custom targets provided via `--target=`, which would allow users to patch the "natural stack alignment" in their custom target which should relax LLVM stack alignment assumptions that are present in the system.
+Another alternative could be adding a new ABI that captures "function which can be called with any stack alignment".
+I chose to propose this RFC and not any of the alternatives because it seems to me that this proposition provides the simplest solution, a solution that is very close to `force_align_arg_pointer` function attribute in GCC and a solution that is easy to implement for rustc.
+While creating a different ABIs to handle stack realignment could be a viable alternative, introducing a new function attribute like realign_stack in Rust offers several advantages. Firstly, leveraging function attributes aligns with Rust's philosophy of providing clear and concise language features, ensuring that developers can easily understand and apply stack realignment where necessary. Also, if the realign_stack was a part of the ABI we would need to practically duplicate every ABI and create a copy where one has that attribute and the other does not. This would lead to a higher level of complexity and would require higher maintenance over time. 

```

Comment:
The key point here is that having the ability to be called with an unaligned stack is orthogonal to Rust's existing ABI definitions; perhaps another approach could be for the language to support ABIs as a composable set of features along different axes, e.g. `extern "C" + "realign_stack" fn ...` or somesuch?  That way, it could still form part of the function's type signature/ABI while avoiding exponential growth in the number of ABIs.

Reply:
this looks good enough to me but im not sure it is much different from something like the `no_mangle` attribute which some people also think should be unsafe.

Reply:
we already have duplicate ABIs for supporting unwinding `extern "C-unwind"`, so I quite like the ABIs as a composable set of features idea! `extern "C" + "unwind" + "align_stack" fn()`

Reply:
I did not know that actually and this does look nice.
However I think a different RFC might be better for that.
Also, should be noted, there is still `#[unwind(allowed)]` and `#[unwind(abort)]` which might suggest that maybe we should be doing both the ABI and attribute solutions